// react-native-draggable/Draggable.js
/**
 *	* https://github.com/tongyy/react-native-draggable
 * 
 */

import React, { Component } from 'react';
import {
	Platform,
	StyleSheet, 
	View,
	Text,
	Image,
	PanResponder,
	Animated,
	Dimensions,
	TouchableOpacity
} from 'react-native';
import PropTypes from 'prop-types';
import * as Animatable from 'react-native-animatable';
import { WaveIndicator } from 'react-native-indicators';




export default class Draggable extends Component {
	static propTypes = {
		renderText:PropTypes.string,
		renderShape:PropTypes.string,
		renderSize:PropTypes.number,
		onPress:PropTypes.func,
		imageSource:PropTypes.oneOfType([
			PropTypes.shape({
				uri: PropTypes.string,
			}),
			PropTypes.number
		]), 
		offsetX:PropTypes.number,
		offsetY:PropTypes.number,
		renderColor:PropTypes.string,
		reverse:PropTypes.bool,
		pressDrag:PropTypes.func,
		onMove:PropTypes.func,
		pressDragRelease:PropTypes.func,
		longPressDrag:PropTypes.func,
		pressInDrag:PropTypes.func,
		pressOutDrag:PropTypes.func,
		z:PropTypes.number,
		x:PropTypes.number,
		y:PropTypes.number
		
	};
	static defaultProps = {
		offsetX : 100,
		renderShape : 'circle',
		renderColor : 'yellowgreen',
		renderText : '＋',
		renderSize : 36,
		offsetY : 100,
		reverse : true
	}

	componentWillMount() {
		if(this.props.reverse == false)
			this.state.pan.addListener((c) => this.state._value = c);
	}
	componentWillUnmount() {
		this.state.pan.removeAllListeners();
	}
	constructor(props, defaultProps) {
		super(props, defaultProps);
		const { pressDragRelease, reverse, onMove } = props;
		this.state = {
			pan:new Animated.ValueXY(), 
			_value:{x: 0, y: 0}
		};

		this.panResponder = PanResponder.create({		
			onMoveShouldSetPanResponder: (evt, gestureState) => true,
			onMoveShouldSetPanResponderCapture: (evt, gestureState) => true,
			onPanResponderGrant: (e, gestureState) => {
				if(reverse == false) {
					this.state.pan.setOffset({x: this.state._value.x, y: this.state._value.y});
					this.state.pan.setValue({x: 0, y: 0});
				}
			},
			onPanResponderMove: Animated.event([null,{ 
				dx:this.state.pan.x,
				dy:this.state.pan.y
			}], {listener: onMove}),
			onPanResponderRelease: (e, gestureState) => {
				if(pressDragRelease)
					pressDragRelease(e, gestureState);
				if(reverse == false)
					this.state.pan.flattenOffset();
				else 
					this.reversePosition();
			} 
		});
	}
	
	_positionCss = () => {
		let Window = Dimensions.get('window');
		const { renderSize, offsetX, offsetY, x, y, z } = this.props;
		return Platform.select({
			ios: {
				zIndex: z != null ? z : 999,
				position: 'absolute',
				top: y != null ? y : (Window.height / 2 - renderSize + offsetY),
				left: x !=null ? x : (Window.width / 2 - renderSize + offsetX)
			},
			android: {
				position: 'absolute',
				width:Window.width,
				height:Window.height,
				top: y != null ? y : (Window.height / 2 - renderSize + offsetY),
				left: x !=null ? x : (Window.width / 2 - renderSize + offsetX)
			},
		});
	}

	_dragItemCss = () => {
		const { renderShape, renderSize, renderColor } = this.props;
		if(renderShape == 'circle') {
			return{
				backgroundColor: renderColor,
				width: renderSize * 2,
				height: renderSize * 2,
				borderRadius: renderSize,
				borderWidth: 2,
    			borderColor: '#ffcd00',
    			shadowColor: '#000',
    			shadowOffset: { width: 0, height: 0 },
    			shadowOpacity: 0.5,
    			shadowRadius: 2,
    			elevation: 2 
			};
		}else if(renderShape == 'square') {
			return{
				backgroundColor: renderColor,
				width: renderSize * 2,
				height: renderSize * 2,
				borderRadius: 0 
			};
		}else if(renderShape == 'image') {
			return{
				width: renderSize,
				height: renderSize,
    			shadowColor: '#000',
    			shadowOffset: { width: 0, height: 0 },
    			position: 'absolute'
			};
		}
	}
	_dragItemTextCss = () => {
		const { renderSize } = this.props;
		return {
			marginTop: renderSize-10,
			marginLeft: 5,
			marginRight: 5,
			textAlign: 'center',
			color: '#fff',
			shadowColor: '#000',
    		shadowOffset: { width: 0, height: 0 },
    		shadowOpacity: 0.5,
    		shadowRadius: 2,
    		elevation: 2 
		};
	}
	_getTextOrImage = () => {
		const { renderSize, renderShape, renderText, imageSource } = this.props;
		if(renderShape == 'image') {
			return(<Image style={this._dragItemCss(renderSize, null, 'image')} source={imageSource}/>);
		}else{
			return (<Text style={this._dragItemTextCss(renderSize)}>{renderText}</Text>);
		}

	}

	reversePosition = () => {
		Animated.spring(						
			this.state.pan,				 
			{toValue:{x:0,y:0}}		 
		).start();
	}

	render() {
		const touchableContent = this._getTextOrImage();
		const { pressDrag, longPressDrag, pressInDrag, pressOutDrag, onPress } = this.props;
		const { renderShape, renderSize, renderColor } = this.props;

		return (
			<Animatable.View 
				style={this._positionCss()}
				animation='bounceInUp'
            	duration={5000}
			>
				<Animated.View 
					{...this.panResponder.panHandlers}
					style={[this.state.pan.getLayout()]}>
					<TouchableOpacity 
						style={{
    						backgroundColor: '#ffcd00',
    						borderWidth: 1,
    						borderColor: '#ffcd00',
    						shadowColor: '#000',
    						shadowOffset: { width: 0, height: 0 },
    						shadowOpacity: 0.5,
    						shadowRadius: 2,
    						elevation: 2,
    						alignItems: 'center',
    						justifyContent: 'center',
    						height: renderSize,
    						width: renderSize,
    						borderRadius: renderSize/2
  						}}
						onPress={onPress}
						onLongPress={longPressDrag}
						onPressOut={pressOutDrag}
					>
						<WaveIndicator animationDuration={3000} size={80}/>
						{touchableContent}	
					</TouchableOpacity>
				</Animated.View>
			</Animatable.View>
		);
	}
}
